### PR TITLE
FLAG-26: Remove logic dependency and add support for new OpenMRS versions

### PR DESCRIPTION
**Pull request by pgesek**
* Removed the logic module dependency
* Added support for building with OpenMRS 1.11 and 2.0
* Did the cleanup mentioned in this talk post: https://talk.openmrs.org/t/status-of-patient-flags-module-and-steps-forward/10033/17

### CODE REVIEW

**Comment by ssmusoke** [comment]
@craigappl Don't all these changes warrant a version bump to 1.4.0-SNAPSHOT too?

**Comment by mogoodrich** [comment]
@ssmusoke  @craigappl  Actually, might even be worth going to 2.0 on the module....

**Comment by pgesek** [comment]
@mogoodrich @ssmusoke I've updated the version to 2.0.0-SNAPSHOT

**Comment by pgesek** [comment]
I also dropped support for 1.7 and 1.8, as discussed on talk

**Comment by ssmusoke** [comment]
@pgesek I am running an SDK version with legacyui 1.3.1 and I run out of memory with the root error below 

`java.lang.ClassNotFoundException: org.openmrs.web.controller.PortletController
	org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1856)
	org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1705)
	org.openmrs.module.ModuleClassLoader.loadClass(ModuleClassLoader.java:563)
	java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	java.lang.ClassLoader.defineClass1(Native Method)
	java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	java.security.AccessController.doPrivileged(Native Method)
	java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	org.openmrs.module.ModuleClassLoader.loadClass(ModuleClassLoader.java:637)
	org.openmrs.module.ModuleClassLoader.loadClass(ModuleClassLoader.java:556)
	java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	org.openmrs.util.OpenmrsClassLoader.loadClass(OpenmrsClassLoader.java:157)
	java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	org.springframework.util.ClassUtils.forName(ClassUtils.java:249)
	org.springframework.beans.factory.support.AbstractBeanDefinition.resolveBeanClass(AbstractBeanDefinition.java:395)
	org.springframework.beans.factory.support.AbstractBeanFactory.doResolveBeanClass(AbstractBeanFactory.java:1349)
	org.springframework.beans.factory.support.AbstractBeanFactory.resolveBeanClass(AbstractBeanFactory.java:1320)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:622)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:591)
	org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1397)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:434)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:404)
	org.springframework.beans.factory.BeanFactoryUtils.beanNamesForTypeIncludingAncestors(BeanFactoryUtils.java:187)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1112)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1038)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:949)
	org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:533)
	org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:331)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1202)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:537)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:476)
	org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:303)
	org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:299)
	org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:962)
	org.springframework.orm.hibernate4.support.OpenSessionInViewFilter.lookupSessionFactory(OpenSessionInViewFilter.java:189)
	org.springframework.orm.hibernate4.support.OpenSessionInViewFilter.lookupSessionFactory(OpenSessionInViewFilter.java:174)
	org.springframework.orm.hibernate4.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:125)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)
	org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)
	org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)
	org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)`

**Comment by ssmusoke** [comment]
@pgesek Just a headsup even with the new changes I am unable to run the module in an reference application 2.7 server let alone mine

**Comment by dkayiwa** [comment]
It would help to give more details in regards to what happens when you try run it.

**Comment by craigappl** [comment]
@ssmusoke, Are you building the module with the appropriate profile? I got it to run on RA 2.6 with mvn clean install -P openmrs-2.0

Once the module runs, I'm able to create priorities and flags. I'm investigating a null pointer exception in the clinician facing dashboard fragment.

**Comment by ssmusoke** [comment]
@pgesek I did not know that I had to compile it with a specific profile. Will try this out again and provide feedback. 

@dkayiwa Is the profile building a consistent feature across other modules, since all I have used to date is `mvn clean install` and it will work fine.

**Comment by craigappl** [comment]
Hi Everyone,

I want to update you for testing purposes. I got past the null pointer exceptions by developing the following configurations:

- Create a tag that is visible to "Provider"
- Create a Priority with style color:red; (don't forget the semicolon, this causes a null pointer exception)
- Create a flag of type SQL flag
criteria: select e.patient_id from encounter e where e.encounter_datetime < now()
Message: Test Flag
Priority: (Select the priority you created)
Associated Tags: (Select the tag you created)

Notes:
- Tags are required, or you'll get a null pointer exception
- Within the tag, the user permissions need to be appropriately allocated to the user, or you won't get the flag to evaluate (this is done in the fragment controller).
- The style directly inserts into an HTML style tag <... style = " INSERTED STYLE " > so any valid CSS is able to be used. I missed the semicolon and I got a controller exception.

Here's a screenshot of a successful red flag:
image

**Comment by pgesek** [comment]
@ssmusoke yes, you need profile i.e. mvn clean install -Popenmrs-2.0 (we support 1.9, 1.10, 1.11 and 2.0). Let me know if there are any issues.

Just a heads up, french translations were added.

**Comment by ssmusoke** [comment]
@pgesek A question - how will this work when deploying the module to different deployments from BinTray via the SDK? How will this work when the module is being watched by UI Framework

**Comment by dkayiwa** [comment]
@pgesek a simple mvn clean install should result into a module that runs on all the platform versions that you target. We do not recommend doing special compilations for omod distributions of different platforms. That is exactly why @ssmusoke is getting that error when he simply does the expected mvn clean install to create an omod.
Therefore, your config.xml should replace "${requireLegacyuiModule}" with this: https://pastebin.com/ap2h8xUR

**Comment by dkayiwa** [comment]
@pgesek are you still working on this?

**Comment by pgesek** [comment]
@dkayiwa apologies - this slipped my mind. I've done the change you asked for in the commit above. Thanks for the explanation.

**Comment by dkayiwa** [comment]
@pgesek can you include a link to the ticket as advised here? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by pgesek** [comment]
@dkayiwa this is a contribution of work we did for Itech. I didn't have a ticket, so I created it: https://issues.openmrs.org/browse/FLAG-26

**Comment by ssmusoke** [comment]
@pgesek I get the error in the gist https://gist.github.com/ssmusoke/741f6dc0d945314acd7af4b330fd8ca0 when I try to select 
    Manage Flags
    Manage Tags
    Find Flagged Patients

I am using Platform 2.0.5

**Comment by dkayiwa** [comment]
@pgesek i just got the same exact error as @ssmusoke

**Comment by ssmusoke** [comment]
@pgesek ping on whether you were able to work on this

**Comment by ssmusoke** [comment]
@pgesek Ping

**Comment by pgesek** [comment]
@ssmusoke apologies, I am a bit swamped recently with other commitments. I've been however able to identify the issue - OpenMRS 1.9 uses Hibernate 3.X while OpenMRS 2.0 uses Hibernate 4. When building with the 2.0 profile, the correct version of Hibernate is used for building and testing and then at runtime with the platfrm. 

The problem comes when you build without the profile - Hibernate 3 is used for building & tests, however at runtime Hibernate 4 is used, hence the error with the missing methods, since Hibernate classes have changed. 

Is there a convention in OpenMRS for dealing with this Hibernate issue? I'll take a look at different modules, I imagine some of them already solved this issue?

**Comment by dkayiwa** [comment]
@pgesek the problem with that approach is that you would have to compile and ship two versions of modules which is discouraged. Our recommendation is here:
https://wiki.openmrs.org/display/docs/Supporting+different+OpenMRS+versions
https://wiki.openmrs.org/display/docs/Supporting+Platform+2.0+and+below

If the above is hard for you to deal with in your limited time, then here is a quick hack 😊 
https://github.com/openmrs/openmrs-module-xforms/commit/5d79f327d27777a40ec755a90433e4dc69067e4a

**Comment by pgesek** [comment]
@dkayiwa @ssmusoke I used the DbSessionFactory and updated the versions to what was mentioned in the wiki page you shared. All test profiles passed and I very briefly tested with OpenMRS 2 and it was working fine (after mvn clean install with no profiles).

**Comment by ssmusoke** [comment]
@pgesek I run into the error below when starting the module 

`WARN - ModuleFactory.startModuleInternal(788) |2017-10-25 04:13:23,552| Error while trying to start module: patientflags
org.openmrs.module.ModuleException: Module requires version matching 1.11.4 - 1.11.*, 1.10.2 - 1.10.*, 1.9.9 - 1.9.*.  Current code version is 2.0.5
	at org.openmrs.module.ModuleUtil.checkRequiredVersion(ModuleUtil.java:355)
	at org.openmrs.module.ModuleFactory.startModuleInternal(ModuleFactory.java:641)
	at org.openmrs.api.context.Daemon$1.run(Daemon.java:74)`

**Comment by pgesek** [comment]
Seems it's because I switched 1.11.4 to 1.11.4 - 1.11.*, and apparently I did this after testing the module locally. Should work with the current ranges: 1.11.4, 1.10.2 - 1.10.*, 1.9.9 - 1.9.*

**Comment by dkayiwa** [comment]
@pgesek your commit will still not work for platform 2.x

**Comment by pgesek** [comment]
Well it works for me, apparently because it compares '1.11.4' to '2.0.5' here: 

https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/module/ModuleUtil.java#L347

which passes. Shouldn't it? Do you want me to change it to 2.0.0, 1.11.4 - 1.11.\*, ... ?

**Comment by dkayiwa** [comment]
@pgesek 1.11.4 - 1.11.* does not apply for anything that does not start with 1.11. So it will not even work for 1.12

When you say it works for you, do you mean that you are able to run this module on platform 2.x after a simple mvn clean install?

File: api/src/test/java/org/openmrs/module/patientflags/FlagTest.java [hunk 1]
```diff
@@ -173,49 +137,23 @@ public void eval_groovyShouldReturnFalseForTestPatient() throws Exception {
...
-	
-	
-	/**
-	 * Mimics org.openmrs.web.Listener.getRuntimeProperties()
-	 * 
-	 * @param webappName name to use when looking up the runtime properties env var or filename
-	 * @return Properties runtime
-	 * @throws Exception 
-	 */
     @SuppressWarnings("deprecation")
```
Comment (ssmusoke) [inline api/src/test/java/org/openmrs/module/patientflags/FlagTest.java]: @pgesek When I run `mvn clean install` I get stuck at running the FlagTest, I suspect the issue may be with the datasets for the different build versions. However the datasets are all the same so why not just import a single XML dataset and delete this function

Comment (dkayiwa) [reply api/src/test/java/org/openmrs/module/patientflags/FlagTest.java]: @pgesek did you see the above? 😊

Comment (ssmusoke) [reply api/src/test/java/org/openmrs/module/patientflags/FlagTest.java]: @pgesek just a ping to bring this back to top of mind

Comment (ssmusoke) [reply api/src/test/java/org/openmrs/module/patientflags/FlagTest.java]: @pgesek Happy Monday, ping on this - thanks

Comment (pgesek) [reply api/src/test/java/org/openmrs/module/patientflags/FlagTest.java]: @ssmusoke I wasn't able to reproduce this hanging issue, perhaps it has something to do with the database connection? I've removed the redundant as you suggested however.

**Comment by pgesek** [comment]
@dkayiwa yes, I built with mvn clean install and it started no problem with OpenMRS 2.0.5. The current required version string is (note that the first part is not a range):

1.11.4, 1.10.2 - 1.10.*, 1.9.9 - 1.9.*

I've debugged OpenMRS and it takes the first version, 1.11.4, then compares it to 2.0.5. Since 2.0.5 is higher than 1.11.4 it allows the module to start. I think that's correct?

**Pull request merged**